glam/api/views.py

```diff
@@ -281,6 +413,105 @@ def get_glean_aggregations(request, **kwargs):
     return response
 
 
+def get_glean_aggregations_from_bq(request, **kwargs):
+    REQUIRED_QUERY_PARAMETERS = [
+        "aggregationLevel",
+        "app_id",
+        "ping_type",
+        "probe",
+        "product",
+    ]
+    if any([k not in kwargs.keys() for k in REQUIRED_QUERY_PARAMETERS]):
+        # Figure out which query parameter is missing.
+        missing = set(REQUIRED_QUERY_PARAMETERS) - set(kwargs.keys())
+        raise ValidationError(
+            "Missing required query parameters: {}".format(", ".join(sorted(missing)))
+        )
+
+    channel = kwargs["app_id"]
+    product = kwargs.get("product")
+    probe = kwargs["probe"]
+    num_versions = kwargs.get("versions", 3)
+    ping_type = kwargs["ping_type"]
+    os = kwargs.get("os", "*")
+    aggregation_level = kwargs["aggregationLevel"]
+
+    table_id = f"glam_{product}_{channel}_aggregates_v1"
+
+    if aggregation_level == "version" and product == "fenix":
+        build_id_filter = 'AND build_id = "*"'
+    else:
+        build_id_filter = 'AND build_id != "*"'
+    # Build the SQL query with parameters
+    query = f"""
+        WITH versions AS (
+            SELECT
+                ARRAY_AGG(DISTINCT version
+                ORDER BY
+                version DESC
+                LIMIT
+                {num_versions}) AS selected_versions
+            FROM
+                `glam_etl.{table_id}`
+            WHERE
+                metric = @metric
+            )
+            SELECT
+            * EXCEPT(selected_versions)
+            FROM
+                `glam_etl.{table_id}`,
+                versions
+            WHERE
+                metric = @metric
+                AND ping_type = @ping_type
+                AND os = @os
+                {build_id_filter}
+                AND version IN UNNEST(versions.selected_versions)
+    """
+    job_config = bigquery.QueryJobConfig(
+        query_parameters=[
+            bigquery.ScalarQueryParameter("metric", "STRING", probe),
+            bigquery.ScalarQueryParameter("ping_type", "STRING", ping_type),
+            bigquery.ScalarQueryParameter("os", "STRING", os),
+        ]
+    )
+    with bigquery.Client(project=GLAM_BQ_PROD_PROJECT) as client:
+        query_job = client.query(query, job_config=job_config)
```

Comment:
Usually, you would use dependency injection to pass a fully instantiated client object rather than create it in the function. A couple of advantages: using the same client object for multiple queries within a route, loose coupling means it's way easier to unit test because you can inject a mocked client, etc.

Reply:
In your case, since PG leverages Django directly and does not require DI, then you could create your BQ Client [here](https://github.com/mozilla/glam/blob/d3919d22afe0414a3f4d2d403715f7cf8ec15e46/glam/api/views.py#L311) and [here](https://github.com/mozilla/glam/blob/d3919d22afe0414a3f4d2d403715f7cf8ec15e46/glam/api/views.py#L56), and pass it as a function argument. That way, your function becomes easier to test.

Reply:
100%
I'll land the currently code and work on this soon